---
title: "Genomicper Notebook"
output: 
  html_document:
    toc: true
  md_document:
      variant: markdown_github
      toc: true
---

## Setup

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook which allows 
code and text to be mixed. If you click on the `Knit` button above RStudio
will generate an html output directory.

Have taken a copy of the R sources of [genomicper](https://github.com/Genomicper/genomicper). 
The routine names (and file names) will have to be renamed but changes/improvements can be made
to be compared with the original data. The `source` command can be used to load the alternative
source files but each function in the file has to be renamed in order not to produce a name
clash. The renaming will take place as/when subroutines are modified. New files should pass
through `lintr` that makes files comply with the [tidyverse style guide](https://style.tidyverse.org/).

```{r setup, echo=FALSE, message=FALSE}

library(genomicper)        # Package to be analysed
library(microbenchmark)    # Microbenchmark routines
library(profvis)           # Profile routines
library(lintr)             # Static R analyser
library(diffr)             # Source file comparison
library(here)              # Determine the root of where the files live

# Make sure the working directory is set to where this file lives.
setwd(paste0(here(),"/Workspaces/Mario"))

```

```{r workflow}

# Remove all objects in the current environment
rm(list=ls())

# Sample workflow from the genomicper documentation

# Load files for analysis
data(demo,SNPsAnnotation)

# Read & format GWAS pvalues
all_data <- read_pvals(data_name=demo,snps_ann=SNPsAnnotation)

# Order data according to the genome
genome_results <-genome_order(all_data=all_data)

# Results from genome_order
ordered_alldata <- genome_results$ordered_alldata
gs_locs <- genome_results$gs_locs

# Create new environment to save variables (e.g. pathways, permutations):
gper.env <- new.env()

# Load more data
data(RHSA164843,RHSA446343,RHSA8876384,RHSA8964572,RHSA109582,RHSA1474244,envir=gper.env)

# Do we need to use environments?
data(RHSA164843,RHSA446343,RHSA8876384,RHSA8964572,RHSA109582,RHSA1474244)


# Map SNPs to pathways 
paths_res <- read2_paths(ordered_alldata=ordered_alldata,
                         gs_locs=gs_locs,sets_from="workspace",sets_prefix="RHSA",
                         level="snp",envir=gper.env)

# Results from read2_paths:		
pers_ids <- paths_res$per_ors
pathways<- paths_res$pathways

```

## Introduction

## Baseline performance

## Pofile

## Questions

* What is the architectural reason to use R environments?

## Possible improvements

- [ ] Add [`roxygen2` documentation](https://cran.r-project.org/web/packages/roxygen2/vignettes/roxygen2.html) to the routines.
- [x] Change `print` to `cat` or `message`, see this [stackoverflow article](https://stackoverflow.com/questions/36699272/why-is-message-a-better-choice-than-print-in-r-for-writing-a-package/36700294).


## Appendix 1: changes to the original code

### read_pvals.R

Check that the file complies with the tidyverse style (no output is good):

```{r read_pvals_lint}
# Check file complies with the tidyverse style using lintr

lint("Src/read_pvals_v2.R")     # Check the file complies with tidyverse style

```

See the output from the modified version of the code:

```{r run_read_pvals_v2}
# Load a modified version
# v2 changes: 
#            print -> cat

source("Src/read_pvals_v2.R")   # Read in routine
all_data <- read_pvals_v2(data_name=demo,snps_ann=SNPsAnnotation)
```

Changes to the original file:

```{r read_pval_source_changes}
# See what changes have been made to the sources
diffr("Src/read_pvals.R","Src/read_pvals_v2.R", width = "100%", height="100%")

```

### genome_order.R

Check that the file complies with the tidyverse style (no output is good):

```{r genome_order_lint}
# Check file complies with the tidyverse style using lintr

lint("Src/genome_order_v2.R")     # Check the file complies with tidyverse style

```

See the output from the modified version of the code:

```{r run_genome_order_v2}
# Load a modified version
# v2 changes: 
#            print -> cat

source("Src/genome_order_v2.R")   # Read in routine

# Order data according to the genome
genome_results <-genome_order_v2(all_data=all_data)

```

Changes to the original file:

```{r genome_order_changes}

# See what changes have been made to the sources
diffr("Src/genome_order.R","Src/genome_order_v2.R", width = "100%", height="100%")

```

### read2_paths.R

Check that the file complies with the tidyverse style (no output is good):

```{r read2_paths_v2_lint}

# Check file complies with the tidyverse style using lintr

lint("Src/read2_paths_v2.R")     # Check the file complies with tidyverse style

```

Example code:

```{r read2_paths_v2}

# Load new variant
source("Src/read2_paths_v2.R")   # Read in routine

# Map SNPs to pathways 
paths_res <- read2_paths_v2(ordered_alldata=ordered_alldata,
                            gs_locs=gs_locs,sets_from="workspace",sets_prefix="RHSA",
                            level="snp",envir=gper.env)

```

Changes to the original file:

```{r read2_paths_changes}

# See what changes have been made to the sources
diffr("Src/read2_paths.R","Src/read2_paths_v2.R", width = "100%", height="100%")

```

### get_pathways.R

Check that the file complies with the tidyverse style (no output is good):

```{r get_pathways_v2_lint}

# Check file complies with the tidyverse style using lintr

lint("Src/get_pathways_v2.R")     # Check the file complies with tidyverse style

```

Changes to the original file:

```{r get_pathways_changes}

# See what changes have been made to the sources
diffr("Src/get_pathways.R","Src/get_pathways_v2.R", width = "100%", height="100%")

```

### hyprbg.R

Check that the file complies with the tidyverse style (no output is good):

```{r hyprbg_v2_lint}

# Check file complies with the tidyverse style using lintr

lint("Src/hyprbg_v2.R")     # Check the file complies with tidyverse style

```

Changes to the original file:

```{r hyprbg_changes}

# See what changes have been made to the sources
diffr("Src/hyprbg.R","Src/hyprbg_v2.R", width = "100%", height="100%")

```
