---
title: "Genomicper Notebook"
output: 
  md_document:
      variant: markdown_github
      toc: true
---

## Setup

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook which allows 
code and text to be mixed. 

```{r setup}
library(genomicper)        # Package to be analysed
library(microbenchmark)    # Microbenchmark routines
library(profvis)           # Profile routines

# Sample workflow

# Load files for analysis
data(demo,SNPsAnnotation)

# Read & format GWAS pvalues
all_data <- read_pvals(data_name=demo,snps_ann=SNPsAnnotation)

# Order data according to the genome
genome_results <-genome_order(all_data=all_data)

# Results from genome_order
ordered_alldata <- genome_results$ordered_alldata
gs_locs <- genome_results$gs_locs
	
```

## Introduction

## Methodology

Have taken a copy of the R sources of [genomicper](https://github.com/Genomicper/genomicper). 
The routine names (and file names) will have to be renamed but changes/improvements can be made
to be compared  with the original data. The `source` command can be used to load the alternative
source files but each function in the file has to be renamed in order not to produce a name
clash. The renaming will take place as/when subrutines are modified.

## Baseline performance

## Pofile

## Possible improvements

- [ ] Add [`roxygen2` documentation](https://cran.r-project.org/web/packages/roxygen2/vignettes/roxygen2.html) to the routines.
- [ ] Change `print` to `cat` or `message`, see this [stackoverflow article](https://stackoverflow.com/questions/36699272/why-is-message-a-better-choice-than-print-in-r-for-writing-a-package/36700294).

## References
